vermouth/citation_parser.py

```diff
@@ -0,0 +1,175 @@
+                              "manual",
+                              "mastersthesis",
+                              "misc",
+                              "phdthesis",
+                              "proceedings",
+                              "techreport",
+                              "unpublished"
+                             ]
+
+    @staticmethod
+    def prepare_file(lines):
+        """
+        Bibtex is not sensetive to line spacing so we join
+        the line as one string. Comment characters are not
+        allowed.
+        """
+        return " ".join(line.strip() for line in lines)
+
+    @staticmethod
+    def find_entries(citation_string):
+        """
+        Look in a string where `@` indicate the
+        begnning of a new entry and return the indices.
+
+        Parameters:
+        -----------
+        citation_string: str
+
+        Yields:
+        --------
+        idx
+            position of '@' in citation_string
+        """
+        for idx, token in enumerate(citation_string):
+            if token == "@":
+                yield idx
+
+    def pop_entry_type(self, entry_string):
+        """
+        Given a string describing a single
+        entry strip that entry from the string
+        and return it. Note the string MUST
+        contain the @.
+
+        Parameters:
+        ------------
+            entry_string: str
+
+        Returns:
+        ---------
+        str, str
+            the entry type and the shortened string
+        """
+        for entry_type in self.known_entries:
+            if entry_string.find(entry_type) == 1:
+                # remove @<entry_type>{
+                entry_string = entry_string[len(entry_type)+1:]
+                return entry_type, entry_string
+        return None
+
+    @staticmethod
+    def pop_key(entry_string):
+        """
+        Given a string of a single entry from which the
+        entry_type has already been removed (see pop_entry_type)
+        get the custom ID, strip it and return the entry_string
+        without that ID.
+
+        Parameters:
+        -----------
+        entry_string: str
+
+        Returns:
+        --------
+        str, str
+            the key and the string without key
+        """
+        key_idx = entry_string.find(',')
+        return entry_string[:key_idx].strip("{").strip(" "), entry_string[key_idx:]
+
+    @staticmethod
+    def extract_fields(entry_string):
+        """
+        Given an entry string without entry type and identified
+        (i.e. ,<field_type> = {<content>}, etc.) split all the
+        contents and field-types using a regular expression.
+
+        Parameters:
+        -----------
+        entry_string: str
+
+        Yields:
+        -------
+        str, str
+            the field type, the field content
+        """
+        for field, value in re.findall("(.*?)=(.*?)\}", entry_string):
+            yield field.strip(",").strip(" "), value.strip("{").strip("}")
```

Comment:
I wonder how many bibtex parsers break if you put an `'='` in your article title... 🤔

Reply:
I wonder to because our parser is fine

Reply:
I missed the trailing } in the regex :)